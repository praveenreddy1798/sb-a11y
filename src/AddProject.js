import React, { useCallback } from 'react';
import {
  Button,
  FormControl,
  FormLabel,
  FormErrorMessage,
  Input,
  HStack,
  Alert,
  AlertIcon,
  CloseButton,
  useDisclosure,
} from '@chakra-ui/react';
import { Formik, Form } from 'formik';

const errorMap = {
  required: 'Please enter a Storybook URL',
  format: 'Please enter a correctly formatted URL',
};

const urlRegExp = new RegExp(
  // eslint-disable-next-line no-useless-escape
  /[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)?/gi
);

const validate = ({ storybookUrl }) => {
  const errors = {};

  if (!storybookUrl) {
    errors.storybookUrl = errorMap.required;
  } else {
    const validStorybookUrl = storybookUrl.match(urlRegExp);

    if (validStorybookUrl === null) {
      errors.storybookUrl = errorMap.format;
    }
  }

  return errors;
};

export const AddProject = () => {
  const {
    isOpen: submitSuccessful,
    onClose: closeSuccessAlert,
    onOpen: openSuccessAlert,
  } = useDisclosure({ defaultIsOpen: false });

  const {
    isOpen: submitFailed,
    onClose: closeFailedAlert,
    onOpen: openFailedAlert,
  } = useDisclosure({ defaultIsOpen: false });

  const handleFormSubmit = useCallback(
    async ({ storybookUrl }, { setSubmitting, resetForm }) => {
      setSubmitting(true);

      fetch('/project', {
        method: 'POST',
        body: JSON.stringify({ storybookUrl }),
      })
        .then((res) => res.json())
        .then(() => {
          setSubmitting(false);
          openSuccessAlert();
          resetForm({
            values: { storybookUrl: '' },
          });
        })
        .catch(() => {
          openFailedAlert();
          setSubmitting(false);
        });
    },
    [openSuccessAlert, openFailedAlert]
  );

  return (
    <>
      <Formik
        initialValues={{
          email: '',
          password: '',
          verifiedPassword: '',
        }}
        onSubmit={handleFormSubmit}
        validate={validate}
      >
        {({
          touched,
          errors,
          values,
          isSubmitting,
          handleChange,
          handleBlur,
        }) => (
          <HStack
            as={Form}
            noValidate
            aria-disabled={isSubmitting ? 'true' : 'false'}
            spacing="3"
            mb={8}
            alignItems="flex-start"
          >
            <FormControl
              id="storybookUrl"
              isRequired
              isInvalid={touched.storybookUrl && errors.storybookUrl}
            >
              <FormLabel srOnly>Storybook URL</FormLabel>
              <Input
                type="text"
                placeholder="Enter the URL for the Storybook"
                size="lg"
                fontSize="md"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.storybookUrl}
                data-testid="storybook-url"
              />
              <FormErrorMessage color="red" mt={2}>
                {errors.storybookUrl}
              </FormErrorMessage>
            </FormControl>
            <Button
              type="submit"
              fontWeight="bold"
              fontSize="md"
              colorScheme="brand"
              size="lg"
              isLoading={isSubmitting}
            >
              Import
            </Button>
          </HStack>
        )}
      </Formik>
      {submitSuccessful && (
        <Alert status="success">
          <AlertIcon />
          Project added successfully
          <CloseButton ml="auto" onClick={closeSuccessAlert} />
        </Alert>
      )}
      {submitFailed && (
        <Alert status="error">
          <AlertIcon />
          Something went wrong. Unable to add this project.
          <CloseButton ml="auto" onClick={closeFailedAlert} />
        </Alert>
      )}
    </>
  );
};
