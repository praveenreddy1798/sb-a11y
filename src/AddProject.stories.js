import React from 'react';
import { rest } from 'msw';
import { within, userEvent, waitFor } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { AddProject } from './AddProject';

export default {
  component: AddProject,
  title: 'Components/AddProject',
  argTypes: {
    onImport: { action: 'onImport' },
  },
};

const Template = (args) => <AddProject {...args} />;

export const Default = Template.bind({});

export const UrlValidationError = Template.bind({});
UrlValidationError.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);

  const urlInput = canvas.getByTestId('storybook-url');
  expect(urlInput).toBeRequired();
  expect(urlInput).toHaveAccessibleName('Storybook URL');
  await userEvent.type(urlInput, 'my-storybook');
  await urlInput.blur();

  await waitFor(() => {
    expect(urlInput).toBeInvalid();
    expect(urlInput).toHaveAccessibleDescription(
      'Please enter a correctly formatted URL'
    );
  });
};

export const SubmitFailed = Template.bind({});
SubmitFailed.parameters = {
  msw: [
    rest.post('/project', (req, res, ctx) => {
      return res(ctx.status(401));
    }),
  ],
};
SubmitFailed.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);

  const urlInput = canvas.getByLabelText(/URL/i);
  await userEvent.type(urlInput, 'my-storybook.com');

  const submitButton = canvas.getByRole('button', { name: /Import/i });
  await userEvent.click(submitButton);
};

export const SubmitSuccess = Template.bind({});
SubmitSuccess.parameters = {
  msw: [
    rest.post('/project', (req, res, ctx) => {
      const project = JSON.parse(req.body);

      return res(
        ctx.delay(1000),
        ctx.json({
          id: 83470,
          url: project.storybookUrl,
        })
      );
    }),
  ],
};
SubmitSuccess.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);

  const urlInput = canvas.getByLabelText(/URL/i);
  await userEvent.type(urlInput, 'my-storybook.com');

  const submitButton = canvas.getByRole('button', { name: /Import/i });
  await userEvent.click(submitButton);
};
