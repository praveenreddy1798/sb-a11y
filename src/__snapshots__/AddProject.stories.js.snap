// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/AddProject Default smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Storybook URL",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "invalid": "true",
      "name": "Storybook URL",
      "pressed": undefined,
      "required": true,
      "role": "textbox",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Import",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "components-addproject--default",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`Components/AddProject SubmitFailed play-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Storybook URL",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Storybook URL",
      "pressed": undefined,
      "required": true,
      "role": "textbox",
      "value": "my-storybook.com",
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Import",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Something went wrong. Unable to add this project.",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Close",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "components-addproject--submit-failed",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`Components/AddProject SubmitSuccess play-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Storybook URL",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Storybook URL",
      "pressed": undefined,
      "required": true,
      "role": "textbox",
      "value": "my-storybook.com",
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Import",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "components-addproject--submit-success",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`Components/AddProject UrlValidationError play-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Storybook URL",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "description": "Please enter a correctly formatted URL",
      "invalid": "true",
      "name": "Storybook URL",
      "pressed": undefined,
      "required": true,
      "role": "textbox",
      "value": "my-storybook",
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Please enter a correctly formatted URL",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Import",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "components-addproject--url-validation-error",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
